### DISCUSSION TITLE
Update Script for poolController

### DISCUSSION THREAD

**Discussion by DrJeff** [tagyoureit/nodejs-poolController#496] [General]
Ok I made a very simple update script you will need to change the locations to where everything was installed on the Pi and also `sudo chmod +x` the script so it can execute. I just named mine `updatePool.sh`  and it's linked below as a Gist but you can name it whatever you like.
What it does is change to the appropriate directory and do a `git pull` if it is not up to date then it will `npm i` after to do the install. 
https://gist.github.com/DrJeff/f20f73f4a30f9ed265b40950fbd005b1
```
#! /bin/bash
# Simple Upgrade Script For njsPC Free To Use And Modify as you see fit 
# @DrJeff
echo updating njsPC, dashpanel, and rem please wait...
echo updating poolController now ...
cd /home/pi/nodejs-poolController
git pull | tail -f | if grep -L "Already up to date." ; then 
npm i
else
echo updating dashPanel ...
fi
cd /home/pi/nodejs-poolController-dashPanel
git pull | tail -f | if grep -L "Already up to date." ; then 
npm i
else
echo updating REM ...
fi
cd /home/pi/relayEquipmentManager
git pull | tail -f | if grep -L "Already up to date." ; then
npm i
else
echo "All up to date, Completed for now!"
fi
```

**Comment by tagyoureit** [comment]
I love the idea, and was thinking about doing something like this.  However, your logic is a bit off.  The current if/then/else will run `npm i` if there are no updates to pull and if there are updates it will echo `updating [product] ...`.

You also don't need the `tail -f`.  eg...
```
if (git pull | grep -q 'Already'); then echo 'No updates to pull'; else echo 'Updates pulled';  fi;
```

Here's my iteration that has the changes in the logic:
```
#! /bin/bash
# Simple Upgrade Script For njsPC Free To Use And Modify as you see fit
# @DrJeff
echo updating njsPC, dashpanel, and rem please wait...

echo updating poolController now ...
cd /home/pi/nodejs-poolController
if (git pull | grep -q "Already up to date.") ; then
echo njsPC already up to date;
else
echo njsPC updates found, installing dependencies.;
npm i;
fi

echo updating dashPanel ...
cd /home/pi/nodejs-poolController-dashPanel
if (git pull | grep -q "Already up to date.") ; then
echo dP already up to date;
else
echo dp updates found, installing dependencies.;
npm i
fi

echo updating REM ...
cd /home/pi/relayEquipmentManager
if (git pull | grep -q "Already up to date.") ; then
echo REM already up to date;
else
echo REM updates found, installing dependencies.;
npm i
fi

echo "All up to date, Completed for now!"
```

**Reply by DrJeff** [reply]
Nah the -L is a not condition or a no-match so it does works properly but I like yours anyways. 😉

**Reply by tagyoureit** [reply]
I couldn't get it working with `-L`.  I tried a few iterations and `grep -v` worked as a no-match, but 🤷‍♂️.  Glad you like the changes.

**Comment by johnny2678** [comment]
Thanks!  Didn't know i needed this but used it this morning and updating was so much easier.  Don't have any self-doubt about did I remember to `npm i` ?? 😂

Added my own pm2 process restarts:

```
#! /bin/bash
# Simple Upgrade Script For njsPC Free To Use And Modify as you see fit
# @DrJeff
echo updating njsPC, dashpanel, and rem please wait...

echo updating poolController now ...
cd /home/pi/nodejs-poolController
if (git pull | grep -q "Already up to date.") ; then
  echo njsPC already up to date;
else
  echo njsPC updates found, installing dependencies.;
  pm2 stop 3
  npm i;
  pm2 start 3
fi

echo updating dashPanel ...
cd /home/pi/nodejs-poolController-dashPanel
if (git pull | grep -q "Already up to date.") ; then
  echo dP already up to date;
else
  echo dp updates found, installing dependencies.;
  pm2 stop 2
  npm i
  pm2 start 2
fi

echo updating REM ...
cd /home/pi/relayEquipmentManager
if (git pull | grep -q "Already up to date.") ; then
  echo REM already up to date;
else
  echo REM updates found, installing dependencies.;
  pm2 stop 1
  npm i
  pm2 start 1
fi

echo "All up to date, Completed for now!"




```

**Reply by tagyoureit** [reply]
That's good, too! 
![2022-06-23_08-19-12](https://user-images.githubusercontent.com/1646840/175334825-21e66a80-eb37-46c5-adce-1c6747de8a67.png)

Did you know you can reference the PM2 processes by name?  If you are using the default ecosystem.config.js, the processes will be named like in the above picture.  There is always a chance you (or someone else) could have different id's so replacing `pm2 stop 3` with `pm2 stop njsPC` would be a little more resilient.

**Reply by DrJeff** [reply]
oops should I be stopping the node prior to install, it always restarts after install. 🤷🏽‍♂️ 
either way I'm gonna steal that as well, I mean use that. 😄

**Reply by tagyoureit** [reply]
> oops should I be stopping the node prior to install, it always restarts after install. 🤷🏽‍♂️

It really won't matter too much.  Once you have the apps running all of the modules are loaded into memory.  PM2 (our default example) won't restart based on the `node_modules` directory changes.  But @johnny2678 is definitely correct in restarting the apps after you run `npm i`.  Otherwise, you won't load all of the great changes you just installed.

**Reply by johnny2678** [reply]
> replacing `pm2 stop 3` with `pm2 stop njsPC` would be a little more resilient.

great point!

**Reply by tagyoureit** [reply]
I noticed that when restarting the watching flag on PM2 isn't reset back to true.  each `pm2 start xyz` should have `--watch` at the end.  
EG `pm2 start njsPC --watch`.

**Reply by johnny2678** [reply]
nice, I'll make sure to add this

**Comment by dsarvadi** [comment]
Is there a version for a windows 10 application?

**Reply by tagyoureit** [reply]
There's not.  But you can just throw the existing scripts into ChatGPT, Gemini, etc and it will convert it for you.

**Comment by dsarvadi** [comment]
ThanksSent from my Galaxy
-------- Original message --------From: tagyoureit ***@***.***> Date: 7/13/26  11:34 AM  (GMT-05:00) To: tagyoureit/nodejs-poolController ***@***.***> Cc: dsarvadi ***@***.***>, Comment ***@***.***> Subject: Re: [tagyoureit/nodejs-poolController] Update Script for poolController (Discussion #496) 
There's not.  But you can just throw the existing scripts into ChatGPT, Gemini, etc and it will convert it for you.
